modules/internal/String.chpl

```diff
@@ -430,31 +433,62 @@ module String {
       }
     }
 
+    /*
+      Iterates over the string Unicode character by Unicode character.
+    */
+    iter uchars(): int(32) {
+      var localThis: string = this.localize();
+
+      var i = 0;
+      while i < localThis.len {
+        var codepoint: int(32);
+        var nbytes: c_int;
+        var multibytes = (localThis.buff + i): c_string;
+        var maxbytes = (localThis.len - i): ssize_t;
+        if maxbytes > 4 then
+          maxbytes = 4;
+        qio_decode_char_buf(codepoint, nbytes, multibytes, maxbytes);
+        yield codepoint;
+        i += nbytes;
+      }
+    }
+
     /*
       Index into a string
 
-      :returns: A string with the character at the specified index from
-                `1..string.length`
+      :returns: A string with the multibyte character at the specified index
+                from `1..string.length`
```

Comment:
This change makes `i` in this function a *byte offset* which is what I think it should be but I'm not sure if other people want to argue about it?  In any case I think it needs to be clear in the docs comment.

Also, what happens if you index halfway through a multibyte character? Is there a test of that? I'd hope you'd get the replacement character `0xfffd` or maybe an error.

Reply:
I'm updating the comment to be more precise.

If you index part way into a multibyte character, it gives you the replacement character.